File: typescript/agent.ts
/**
 * Workflow processor agent - TypeScript example.
 *
 * Listens for intents via SSE, processes order fulfillment workflows,
 * resumes with completion results.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   npx tsx agent.ts
 */

import { AxmeClient } from "@axme/axme";

const AGENT_ADDRESS = "stepfn-alt-processor-demo";

async function handleIntent(client: AxmeClient, intentId: string) {
  const intentData = await client.getIntent(intentId);
  const intent = intentData.intent ?? intentData;
  let payload = intent.payload ?? {};
  if (payload.parent_payload) {
    payload = payload.parent_payload;
  }

  const workflowId = payload.workflow_id ?? "unknown";
  const workflowType = payload.workflow_type ?? "unknown";
  const orderId = payload.order_id ?? "unknown";
  const steps: string[] = payload.steps ?? [];

  console.log(`  Workflow: ${workflowId}`);
  console.log(`  Type: ${workflowType}`);
  console.log(`  Order: ${orderId}`);
  console.log(`  Steps: ${steps.length}`);

  // Execute each step
  for (let i = 0; i < steps.length; i++) {
    console.log(`  [${i + 1}/${steps.length}] ${steps[i]}...`);
    await new Promise((r) => setTimeout(r, 1000));
  }

  const result = {
    action: "complete",
    workflow_id: workflowId,
    order_id: orderId,
    steps_completed: steps.length,
    order_status: "fulfilled",
    completed_at: new Date().toISOString(),
  };

  await client.resumeIntent(intentId, result, { ownerAgent: AGENT_ADDRESS });
  console.log(`  Workflow ${workflowId} completed. Order status: ${result.order_status}`);
}

async function main() {
  const apiKey = process.env.AXME_API_KEY;
  if (!apiKey) {
    console.error("Error: AXME_API_KEY not set.");
    process.exit(1);
  }

  const client = new AxmeClient({ apiKey });

  console.log(`Agent listening on ${AGENT_ADDRESS}...`);
  console.log("Waiting for intents (Ctrl+C to stop)\n");

  for await (const delivery of client.listen(AGENT_ADDRESS)) {
    const intentId = delivery.intent_id;
    const status = delivery.status;

    if (!intentId) continue;

    if (["DELIVERED", "CREATED", "IN_PROGRESS"].includes(status)) {
      console.log(`[${status}] Intent received: ${intentId}`);
      try {
        await handleIntent(client, intentId);
      } catch (e) {
        console.error(`  Error processing intent: ${e}`);
      }
    }
  }
}

main().catch(console.error);

File: agent.py
"""
Workflow processor agent - processes order fulfillment workflows.

Listens for intents via SSE. Simulates a 4-step order fulfillment
workflow and resumes with completion results.

Usage:
    export AXME_API_KEY="<agent-key>"
    python agent.py
"""

import os
import sys
import time

sys.stdout.reconfigure(line_buffering=True)

from axme import AxmeClient, AxmeClientConfig


AGENT_ADDRESS = "stepfn-alt-processor-demo"


def handle_intent(client, intent_id):
    """Process order fulfillment workflow - 4 steps."""
    intent_data = client.get_intent(intent_id)
    intent = intent_data.get("intent", intent_data)
    payload = intent.get("payload", {})
    if "parent_payload" in payload:
        payload = payload["parent_payload"]

    workflow_id = payload.get("workflow_id", "unknown")
    workflow_type = payload.get("workflow_type", "unknown")
    order_id = payload.get("order_id", "unknown")
    steps = payload.get("steps", [])

    print(f"  Workflow: {workflow_id}")
    print(f"  Type: {workflow_type}")
    print(f"  Order: {order_id}")
    print(f"  Steps: {len(steps)}")

    # Execute each step
    for i, step in enumerate(steps, 1):
        print(f"  [{i}/{len(steps)}] {step}...")
        time.sleep(1)

    result = {
        "action": "complete",
        "workflow_id": workflow_id,
        "order_id": order_id,
        "steps_completed": len(steps),
        "order_status": "fulfilled",
        "completed_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    }

    client.resume_intent(intent_id, result)
    print(f"  Workflow {workflow_id} completed. Order status: {result['order_status']}")


def main():
    api_key = os.environ.get("AXME_API_KEY", "")
    if not api_key:
        print("Error: AXME_API_KEY not set.")
        print("Run the scenario first: axme scenarios apply scenario.json")
        print("Then get the agent key from ~/.config/axme/scenario-agents.json")
        sys.exit(1)

    client = AxmeClient(AxmeClientConfig(api_key=api_key))

    print(f"Agent listening on {AGENT_ADDRESS}...")
    print("Waiting for intents (Ctrl+C to stop)\n")

    for delivery in client.listen(AGENT_ADDRESS):
        intent_id = delivery.get("intent_id", "")
        status = delivery.get("status", "")

        if not intent_id:
            continue

        if status in ("DELIVERED", "CREATED", "IN_PROGRESS"):
            print(f"[{status}] Intent received: {intent_id}")
            try:
                handle_intent(client, intent_id)
            except Exception as e:
                print(f"  Error processing intent: {e}")


if __name__ == "__main__":
    main()
